src/modules/content-s3/content-s3.c

```diff
@@ -0,0 +1,287 @@
+
+    if (flux_request_decode_raw (msg, NULL, &data, &size) < 0)
+        goto error;
+    if (blobref_hash (ctx->hashfun,
+                      (uint8_t *)data,
+                      size,
+                      blobref,
+                      sizeof (blobref)) < 0)
+        goto error;
+    if (s3_put (ctx->cfg, blobref, data, size, &errstr) < 0)
+        goto error;
+    if (flux_respond_raw (h, msg, blobref, strlen (blobref) + 1) < 0)
+        flux_log_error (h, "error responding to store request");
+    return;
+
+error:
+    if (errstr)
+        printf("\n\n%s\n\n", errstr);
+    if (flux_respond_error (h, msg, errno, errstr) < 0)
+        flux_log_error (h, "error responding to store request");
+}
+
+/* Handle a kvs-checkpoint.get request from the rank 0 kvs module.
+ * The KVS stores its last root reference here for restart purposes.
+ *
+ * N.B. filedb_get() calls read_all() which ensures that the returned buffer
+ * is padded with an extra NULL not included in the returned length,
+ * so it is safe to use the result as a string argument in flux_respond_pack().
+ */
+void checkpoint_get_cb (flux_t *h, flux_msg_handler_t *mh, const flux_msg_t *msg, void *arg)
+{
+    const char *errstr = NULL;
+    struct content_s3 *ctx = arg;
+    const char *key;
+    void *data = NULL;
+    char *dup = NULL;
+    size_t size;
+
+    if (flux_request_unpack (msg, NULL, "{s:s}", "key", &key) < 0)
+        goto error;
+
+    if (s3_get (ctx->cfg, key, &data, &size, &errstr) < 0)
+        goto error;
+
+    if (!(dup = strndup (data, size)))
+        goto error;
+
+    if (flux_respond_pack (h,
+                           msg,
+                           "{s:s}",
+                           "value",
+                           size > 0 ? dup : "", 0) < 0) {
+        errno = EIO;
+        flux_log_error (h, "error responding to kvs-checkpoint.get request (pack)");
+    }
+    free (data);
+    free (dup);
+    return;
+
+error:
+    if (errstr)
+        printf("\n\n%s\n\n", errstr);
+    if (flux_respond_error (h, msg, errno, errstr) < 0)
+        flux_log_error (h, "error responding to kvs-checkpoint.get request");
+    free (data);
+    free (dup);
+}
+
+/* Handle a kvs-checkpoint.put request from the rank 0 kvs module.
+ * The KVS stores its last root reference here for restart purposes.
+ */
+void checkpoint_put_cb (flux_t *h, flux_msg_handler_t *mh, const flux_msg_t *msg, void *arg)
+{
+    struct content_s3 *ctx = arg;
+    const char *key;
+    const char *value;
+    const char *errstr = NULL;
+
+    if (flux_request_unpack (msg,
+                             NULL,
+                             "{s:s s:s}",
+                             "key",
+                             &key,
+                             "value",
+                             &value) < 0)
+        goto error;
+    if (s3_put (ctx->cfg, key, value, strlen (value), &errstr) < 0)
+        goto error;
+    if (flux_respond (h, msg, NULL) < 0) {
+        errno = EIO;
```

Comment:
since `flux_respond()` sets errno on errro, this is unnecessary (it overwites the actual error in fact).